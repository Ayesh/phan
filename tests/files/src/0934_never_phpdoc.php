<?php
namespace NS934;

/**
 * @param mixed $value
 * @return never
 */
function shouldReturnNever($value) {
    // these are all invalid
    if (rand(0, 1)) {
        return $value;
    } elseif ($value) {
        return;
    }
    return null;
}

/**
 * @return never this is valid
 */
function neverExceptional(): void {
    throw new \RuntimeException();
}

/**
 * @param mixed $value
 * @return void
 */
function shouldReturnVoid($value) {
    return $value;  // should warn
}
