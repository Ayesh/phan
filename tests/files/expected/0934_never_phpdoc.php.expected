%s:11 PhanTypeMismatchReturn Returning $value of type mixed but shouldReturnNever() is declared to return never
%s:13 PhanTypeMismatchReturnProbablyReal Returning void of type void but shouldReturnNever() is declared to return never (no real type) (the inferred real return type has nothing in common with the declared phpdoc return type)
%s:15 PhanTypeMismatchReturnProbablyReal Returning null of type null but shouldReturnNever() is declared to return never (no real type) (the inferred real return type has nothing in common with the declared phpdoc return type)
%s:30 PhanTypeMismatchReturn Returning $value of type mixed but shouldReturnVoid() is declared to return void
